File: exercises/euler/problem92.clj
(use 'commons)

(defn next-square-chain [number]
  (let [d (digits-from number)]
    (reduce + (map pow2 d))))

(defn arrives [number]
  (loop [n number]
    (if (or (= n 1) (= n 89))
      n
      (recur (next-square-chain n)))))

(println (arrives 44))
(println (arrives 85))

File: exercises/euler/problem92.rb
upper = 100000

def next_square_chain(number)
  digits = number.to_s.chars.map(&:to_i)
  return digits.map { |d| d*d }.reduce(&:+)
end

arrives = []
arrives[1] = 1
arrives[89] = 89

(1..upper).each do |i|
  chain = [i]
  while arrives[chain.last].nil?
    chain << next_square_chain(chain.last)
  end
  a = arrives[chain.last]
  chain.each do |c|
    arrives[c] = a
  end
end

puts arrives[0..upper].count { |n| n == 89 }
